### ISSUE TITLE
add CDEPS as a library with all data components included

### ISSUE THREAD

**Issue by junwang-noaa** [ufs-community/ufs-weather-model#616]
## Description
This PR is to add CDEPS as a library with all data components included in UFS. Currently the data component (cdeps:datm) is treated as a single data model in UFS. 

## Solution
Build CDEPS as a library which contains other data model components, those data model can be called when it is requested in the run sequence. 

## Related PRs:

- [ufs-weather-model PR#611](https://github.com/ufs-community/ufs-weather-model/pull/611)

**Comment by junwang-noaa** [comment]
@ uturuncoglu I thought you have the patch and it is working for you, is it?

Also can you explain why we need DICE and DOCN in NEMS at this time? These can be added later if they are requested in the run sequence as a model component. But we should still be able to compile the code and run CDEPS DATM tests without those updates. 

@binli2337 can you check why the DATM test is still working while the attributes do not show in the nems.configure file? ( https://github.com/aerorahul/ufs-weather-model/blob/feature/cmake_flags_cleanup/tests/parm/nems.configure.datm_cdeps.IN)

**Comment by aerorahul** [comment]
There are 2 lines in the top-level `CMakeLists.txt`:
```
  list(APPEND _ufs_defs_private CDEPS-interface/CDEPS
                                FRONT_CDEPS_DATM=atm_comp_nuopc)
  include_directories(${CMAKE_CURRENT_BINARY_DIR}/CDEPS-interface/CDEPS/datm)
```

I am not sure how this `CDEPS-interface/CDEPS` got added to `_ufs_defs_private`?
Also, this line is not necessary and unclear how it got added past review.
```
include_directories(${CMAKE_CURRENT_BINARY_DIR}/CDEPS-interface/CDEPS/datm)
```

**Comment by uturuncoglu** [comment]
@junwang-noaa no, I have no patch for this issue. I need to add only DOCN in HAFS level. Since there is no control in build to enable data components individually it will still create problem in UFS level. I could add docn as following,

```
if(CDEPS)  
add_dependencies(ufs cdeps::datm)  
list(APPEND _ufs_defs_private CDEPS-interface/CDEPS                                
      FRONT_CDEPS_DATM=atm_comp_nuopc)  
include_directories(${CMAKE_CURRENT_BINARY_DIR}/CDEPS-interface/CDEPS/datm)  
target_link_libraries(ufs PUBLIC cdeps::datm)

add_dependencies(ufs cdeps::docn)  
list(APPEND _ufs_defs_private CDEPS-interface/CDEPS                                
      FRONT_CDEPS_DATM=ocn_comp_nuopc)  
include_directories(${CMAKE_CURRENT_BINARY_DIR}/CDEPS-interface/CDEPS/docn)  
target_link_libraries(ufs PUBLIC cdeps::docn)
endif()
```

but you will still have still conflict when you build the model with active ocean component if it has same module with data component such as ocn_comp_nuopc. We got error from DICE and CICE because they are using same module and build picks the DICE one and try to configure the model with data ice rather then active ice component. That could be a problem also for the other components. I overcomed this issue by adding support to enable individual data components in build system, which I think it is best way to go and you don't need to worry about the rest of the model. You just need to define application and prevent to use both active components and its data one together. If you really want to use for example FV3 with data atmospeger together then you need to use CDEPS inline data component feature.

**Comment by junwang-noaa** [comment]
@uturuncoglu It looks to me the conflict is caused by the same module name "ice_comp_nuopc" used by both DICE and CICE. This is not the case for other components such as fv3/datm or mom6/docn. I do not see significant difference from your approach of building individual data components and the approach of building cdeps lib with dice included. Actually I am worried that different compile order/different compiler may cause issue even in your approach as we've seen in the past about same module names. I think a clear solution is to use a unique module name for each model, rather than using "ice_comp_nuopc", which could be any ice model, can we use "cdeps_dice_comp_nuopc" to represent CDEPS ice data model? This way we can avoid other model to call the wrong module file existing in the same executable. The same for other CDEPS data components.

**Comment by uturuncoglu** [comment]
@junwang-noaa if you won't compile active component with its data counterpart in the same time and if you allow enabling individual data components that will solve the problem quickly. Otherwise, you need to have PR in CDEPS, CMEPS etc. to reflect the changes in the data components (I mean renaming modules) and all of them need to be tested with CESM etc. (probably require lots of change in there also) and take time. I think we (at least me) don't have that couple of weeks to make it like that before starting to work in HAFS level.

**Comment by rsdunlapiv** [comment]
Longer term, I would advocate for updating the CDEPS module names to be unique.  That seems to be the right approach in the long run because it provides guarantees that the user is not accidentally getting a data model when she intends to get an active model and vice versa.  In other words, we should not be depending on the linking order to make these decisions.

I agree with @uturuncoglu that it will take some coordinated effort to get that set of PRs through the queues.  However, I think the changes themselves are small and straightforward.

@uturuncoglu is your HAFS PR with DOCN ready to go in?  If so, why can't we merge that first and then take on the broader task of renaming the CDEPS modules in a later PR?

**Comment by junwang-noaa** [comment]
@uturuncoglu The issue is that your approach may still fail on other platforms or other compiler. Can you explain why the compiler picks the right module when you compile with individual data components while not pick up the right one in CDEPS lib. We had similar issue with Mersenne mind twister (see ufs-weather-model issue[ #324](https://github.com/ufs-community/ufs-weather-model/issues/324)). Your approach could be a temporary fix, but it does not resolve the issue of same module names. 

Also it looks to me you do have a working version, would that help you start the work in HAFS?

**Comment by junwang-noaa** [comment]
Also another way is to do increment updates. In the first step, we do not include dice in Rahul's PR since only datm and docn are used at this time, please let us know if this is not correct. Once it is committed, it will work with HAFS and we can have a second PR to resolve the name issue.

**Comment by uturuncoglu** [comment]
@rsdunlapiv no it is not ready. there was another issue except this one and also HAFS level RTs needs to be updated to reflect latest changes in HYCOM side and also changes in the UFS build. Also, I still don't have access to Orion.

**Comment by uturuncoglu** [comment]
@junwang-noaa if you don't compile active component such as CICE6 with DICE, it won't fail because you won't have same module. Of course, this is true if you enable adding individual data components. If you merge all data components libraries together to create libcdeps.a (like in the PR) and add it to the link, then you will also have libdice.a in it (which you don't need because you are trying to create configuration with active sea-ice component) and it will fail.

**Comment by uturuncoglu** [comment]
@junwang-noaa the same module names will only create issue if you compile them or link them. Otherwise, it will just work fine. In long term, if you want to use data components also with their active ones such as CICE 6 with DICE together. Then, you need to have a better solution but at this point there is no such configuration or attempt to it. Also, you could still use inline data component if this is the case.

**Comment by junwang-noaa** [comment]
@uturuncoglu you just confirmed that we only need to remove dice section and not add TARGET_OBJECTS:dice to CDEPS in CDEPS-interface/CMakeLists.txt,. If this works we won't need a temporary fix that we don't know if it will work on all the platforms. I think It's much simpler than specifying all the data component at compile time.

**Comment by uturuncoglu** [comment]
@junwang-noaa okay. Let's assumed that we removed dice and we solved the issue with cice but what happens when I try to compile data atmosphere with hycom ocean? It will again fail due to the conflict and both hycom and docn will have same modules. The problem in the existing approach you always need to link docn library along with datm because there is no any information about requested data components. In this case I only need datm library but build also provides docn.

**Comment by binli2337** [comment]
> @binli2337 can you check why the DATM test is still working while the attributes do not show in the nems.configure file? ( https://github.com/aerorahul/ufs-weather-model/blob/feature/cmake_flags_cleanup/tests/parm/nems.configure.datm_cdeps.IN)

@junwang-noaa The datm test is working because the following attributes are included in nems.configure.datm_cdeps.IN for atm component:

diro = “.”
logfile = atm.log

**Comment by uturuncoglu** [comment]
@binli2337 i am confused that the datm configuration is working when you provide 

```
diro = “.”
logfile = atm.log
```

Currently NEMS driver does not read `*_modelio::` attributes, which is a special attribute section used under CESM and they won't be available to CDEPS. So, the run must fail unless you add them to the `ALLCOMP_attributes::` or `ATM_attributes::` section. Is this what you did? I am just asking to clarify it. Anyway, I add section to nems to read `*_modelio::` and it is in the following commit https://github.com/hafs-community/NEMS/commit/279746ab838119098f167a07c5356c5b02ba5a0d

**Comment by junwang-noaa** [comment]
The code was committed in PR#642, close issue.
